docs: Add a proposal to reduce Kiagnose's permissions (#147)
kiagnose/kiagnose: docs/design-proposals/permissions_reduction_proposal.md

Permissions Reduction Proposal
=

Author: [Orel Misan](https://github.com/orelmisan)

> # Summary

Reviewer: Please consider moving the details to the Motivation/Overview section, in a way the reader could understand the problem and the proposed solution in a few sentences, I added a few suggestions.

Currently, kiagnose has a feature to execute a checkup in a dedicated
ephemeral [Namespace](https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces/).
This causes Kiagnose to require very high, cluster-wide permissions for its operation. For example:

1. Get, List, Create, delete and Watch `Namespace` objects.
2. Get, List, Create and Delete `ServiceAccount` objects.
3. Get, List, Create, Delete and Bind `ClusterRole` and `Role` objects.

The proposed solution is:

> 1. To use a user-supplied pre-configured ServiceAccount, to execute a checkup with.

Reviewer 2: ```suggestion
1. To use a user-supplied pre-configured ServiceAccount, with which the checkup will be executed.
```

Author: Done.

> 2. Remove the ability to execute checkups in an ephemeral namespaces.

Reviewer 2: ```suggestion
2. Remove the ability to execute checkups inside ephemeral namespaces.
```

Author: Done.

> The proposed solution is:
> 
> 1. To use a user-supplied pre-configured ServiceAccount, with which the checkup will be executed.
> 2. Remove the ability to execute checkups inside ephemeral namespaces.
> 
> The proposed change will:
> 
> 1. Limit checkup execution to existing namespaces.
> 2. Move the responsibility of creating the proper `ServiceAccount` object for the checkup to the user.
> 3. Remove the majority of the RBAC-related permissions Kiagnose has.

Reviewer: ```suggestion
The proposed solution is to reduce the permissions required for Kiagnose to operate by moving the responsibility of creating the checkup ServiceAccount and its permissions to the user.
```

# Motivation

## Goal

1. Reduce the permissions Kiagnose requires for its operation.
2. Reduce the permissions checkups require for their operation.

> 3. Remove the ephemeral namespace feature.

Reviewer: I am not sure it's a goal, it's a requirement in order to implement (1) and (2).

> # Overview

Reviewer: I am a little bit confused, it seems that this section describes the proposed solution instead of providing background about the current state and operation flow.
I think the details about the proposal should go to Proposal section (maybe after User Stories).

> ## Kiagnose Deployment

Reviewer 3: Is any of this different from the current state?

Author: There is a mistake in part.
Kiagnose deployment should stay the same.
Thanks.

Author: It will just be granted fewer privileges.

Kiagnose shall be deployed to an existing Namespace.

> Kiagnose deployment shall include the following objects:

Reviewer 2: not sure I understand, AFAIR deployment object does not have these fields

Author: ```suggestion
Kiagnose deployment manifest shall include the following objects:
```
The meaning was to [this](https://github.com/kiagnose/kiagnose/blob/main/manifests/kiagnose.yaml) file, and not to K8s Deployment object.

Author: Fixed the wording.

> ## Kiagnose Deployment
> 
> Kiagnose shall be deployed to an existing Namespace.
> 
> Kiagnose deployment manifest shall include the following objects:
> 
> - ServiceAccount
> - ClusterRole
> - ClusterRoleBinding
> 

Reviewer: ```suggestion
Kiagnose shall be deployed to an existing Namespace, including:
- ServiceAccount
- ClusterRole
- ClusterRoleBinding
```

Author: This whole section was a mistake copied from a non-relevant proposal.

> ## Checkups Deployment

Reviewer 3: The title is confusing since this text below is not deploying a checkup. Perhaps "checkup execution pre-requirements"?

A checkup shall be deployed to an existing Namespace.

> A checkup deployment shall include the following objects:

Reviewer 2: ditto

Author: Same answer as in the [previous comment](https://github.com/kiagnose/kiagnose/pull/147#discussion_r946448147).

Author: Fixed the wording.

> ## Checkups Deployment
> 
> A checkup shall be deployed to an existing Namespace.
> 
> A checkup deployment manifest shall include the following objects:
> 
> - ServiceAccount.
> - Role (optional).
> - RoleBinding (optional).
> - ClusterRole (optional).
> - ClusterRoleBinding (optional).
> 
> The user who deploys these objects is responsible for reviewing the permissions given to the checkup.
> 

Reviewer: ```suggestion
A checkup shall be deployed to an existing Namespace, including:
- ServiceAccount.
- Role (optional).
- RoleBinding (optional).
- ClusterRole (optional).
- ClusterRoleBinding (optional).
```

> ## Checkup Configuration
> 
> Checkup configuration shall include the name of the pre-deployed checkup’s ServiceAccount object.
> 

Reviewer: ```suggestion
The user should provide a ServiceAccount granted with the necessary permissions the checkup requires.
The Checkup configuration `ConfigMap` shall include the name of the user-provided ServiceAccount name.
```

## Checkup Setup

The Kiagnose Job shall be created in the `kiagnose` namespace.
The Kiagnose Job shall create the following objects in the target namespace:

1. Results ConfigMap.

> 2. Role and RoleBinding to the user-supplied `ServiceAccount` so the checkup could write its results to the

Reviewer 3: Any reason to make an exception about "supplied service account"? If we required provider of the checkup to document role and rolebinding for a config map as a part of they spec, we remove all management of RBAC.

Author: We can move the responsibility of creating the results ConfigMap to the checkups.
It will indeed remove all RBAC responsibility from Kiagnose, but will require all checkups to have an additional permission.

> The Kiagnose Job shall be created in the `kiagnose` namespace.
> The Kiagnose Job shall create the following objects in the target namespace:
> 
> 1. Results ConfigMap.
> 2. Role and RoleBinding to the user-supplied `ServiceAccount` so the checkup could write its results to the
>    results `ConfigMap`.
> 3. Checkup Job connected to the pre-deployed checkup’s ServiceAccount object.

Reviewer: ```suggestion
The Kiagnose Job shall be created in the `kiagnose` namespace and create the following objects in the target namespace:

1. Results ConfigMap.
2. Role and RoleBinding to the user-supplied `ServiceAccount` so the checkup could write its results to the
   results `ConfigMap`.
3. Checkup Job associated with user-provided ServiceAccount.
```

> ## Checkup Execution
> 
> Without a change.
> 

Reviewer: ```suggestion
The checkup execution flow is not changed.
```

## Checkup Teardown

> The Kiagnose Job shall delete all objects it created during its execution:

Reviewer 2: what about ownerRefenrence? isn't that doable now?

Author: We still need to create and remove a Role (and RoleBinding) to enable the checkup to update its results ConfigMap.
So this option is still not applicable.

> ## Checkup Teardown
> 
> The Kiagnose Job shall delete all objects it created during its execution:
> 
> - Job
> - RoleBinding
> - Role
> - Results ConfigMap
> 
> Kiagnose shall not delete the user-supplied `ServiceAccount` or any of its user-supplied permissions.

Reviewer: ```suggestion
Kiagnose shall delete only the objects it created during execution:
- Job
- RoleBinding
- Role
- Results ConfigMap

Kiagnose shall not delete the user-supplied `ServiceAccount` or any of its user-supplied permissions.
```

# Proposal

## Definition of Personas

> - Kubernetes cluster administrators.

Author: No, because this proposal is only applicable to users that are able to manually trigger a Job in the `kiagnose` namespace.

## User Stories

As a K8s cluster administrator:

> - I want to deploy Kiagnose with less powerful cluster-wide privileges.

Author: > Maybe "I want to deploy Kiagnose without cluster-admin previliges"

Even after the proposed change, in order to deploy `kiagnose` you still need to have the ability to:
1. Create a namespace (`kiagnose`).
2. Create a ServiceAccount in that namespace.
3. Create a ClusterRole and a ClusterRoleBinding.

The "thing" that is less powerful is the ClusterRole, it has less privileges.
Pleas see PR #150 for more details.

- I want to deploy checkups with less powerful privileges to specific namespaces.
- I want to create a checkup configuration, in the target namespace, in order to configure the checkup.
- I want to create a Kiagnose Job in the `kiagnose` namespace, so Kiagnose shall execute the checkup.

> - I want to be able to monitor and observe if the checkup is undergoing or finished.

Author: Yes, it is taken from the [design doc](https://github.com/kiagnose/kiagnose/blob/main/docs/kiagnose_design.md) IIRC.

- I need to manually verify that the checkup had cleaned after itself.

## API Extensions

### User Interface

The following fields shall be included in the user-supplied `ConfigMap` object:

| Key                     | Description                                                              | Is Mandatory | Remarks                                      |
|-------------------------|--------------------------------------------------------------------------|--------------|----------------------------------------------|
| spec.image              | Where to pull the checkup's image from                                   | True         | A registry accessible to your cluster        |
| spec.timeout            | After how much time should Kiagnose stop the running checkup             | True         | 5m, 1h etc                                   |

> | spec.serviceAccountName | Existing serviceAccount for the checkup’s Job                            | True         | `default` ServiceAccount name is not allowed |

Reviewer 2: why is default not allowed?

Author: Kiagnose has to give the input ServiceAccount another permission - the ability to update the results ConfigMap.
If we'll allow the `default` ServiceAccount as an input, every Pod that will be created in that namespace (without a ServiceAccount or with the default) will have that extra permission.

Reviewer 2: So Kiagnose is mutating the input ServiceAccount? isn't that a bit intrusive?

Author: This input ServiceAccount is supposed to be dedicated to a specific checkup.
Kiagnose will not mutate the ServiceAccount, it will bind one Role to it, and remove the binding at the teardown stage.

Reviewer 2: I think it's better than I originally thought, but just to understand, why aren't we giving this responsibility to the ServiceAccount creator (i.e. the sysAdmin)?

Author: Currently checkups write their results to the results ConfigMap which is created by Kiagnose.
Checkups are provided with the namespace and name of this results ConfigMap as an input (using env vars).
In order for the checkups to be able to write to this ConfigMap, Kiagnose creates a Role with write access to a single results ConfigMap (by name).

If the user will be responsible for creating a results ConfigMap and the accompanying Role and RoleBinding, IMO it will:
1. Affect two consecutive (or parallel) runs of the checkup - the results ConfigMap will contain data from a previous run.
2. Add more manual labor on the user.

I think Kiagnose should automate every action that is not privileged and requires a human decision.

Author: > I think it's worth adding a note about default service accounts.

Do you have a suggestion for this note's phrasing?

Reviewer: Similar to what you mentioned above, something like:

> When the default service account is being used for the checkup it will be granted permission to update the result configmap.
> New pods that will be created with the default SA (if no other was specified) will be able to update the checkup result configmap and potentially cause unexpected results.

Author: The "default" ServiceAccount name is an illegal input value in order to avoid this problem. 
A [link](https://github.com/kiagnose/kiagnose/blob/2c517160510ff406a95494b65cc622818441849e/kiagnose/internal/config/cmparser.go#L114) to the implementation.

| spec.spec.param.*       | Arbitrary strings that will be passed to the checkup as input parameters | False        | [0..N]                                |

> ## Implementation Steps

Reviewer 2: I am missing the mechanism to prevent us from hijacking namespaces, we should run checkups only on namespaces that have a "kiagnose-enabled" label

Author: It is contained in PR #126 as an optional feature.
I'm in favor of implementing it, but I don't think it belongs in this document.
